### ISSUE TITLE
apa_print for lmer output

### ISSUE THREAD

**Issue by mcfrank** [crsh/papaja#154]
Thanks for all the work on papaja! It would be wonderful if you could `apa_print` merMods, since these models are the standard for much work in psycholinguistics these days.

**Comment by crsh** [comment]
Hi Mike,

thanks for reaching out. Support for hierarchical models is on the top of my wishlist. I'm currently working on methods for `BFBayesFactor` and related objects as well as `lsmobj` and `ref.grid` objects from the `lsmeans` package. Once, those are done I'll start on support for of `lme4` and mixed model objects from `afex`.

If I may, I'll respond to this thread again once I start working on this to get your opinion on the information `apa_print()` should return.

**Comment by mcfrank** [comment]
Thanks, Frederik. I'd be happy to brainstorm what the output should look
like!

**Comment by jupitus** [comment]
Very pleased to see the `BFBayesFactor` table output--thanks! Any update on `lme4` model objects?

**Comment by crsh** [comment]
There are a couple more loose ends to tie up. I'd like to get the first version of `papaja` to CRAN before starting this new feature. I hope, I'll get going on this no later than at the [SIPS 2019 conference](https://www.improvingpsych.org/SIPS2019/). I'm half-planning an ad-hoc hackathon to work on this.

The first and most important question is what information needs to be available for reporting and how it should be formatted (e.g., tables). What a current best practices? Any thoughts, suggestions, and examples are welcome.

**Comment by khusmann** [comment]
I'm very interested in this too! Here's a possible template that I use when reporting results: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2975938/pdf/nihms245694.pdf (see table 2)

**Comment by crsh** [comment]
Great, thank you.

To add to this, here are the papers I know that come closest to "best practice" guidelines:

- [Bolker et al. (2009)](http://max2.ese.u-psud.fr/epc/conservation/PDFs/HIPE/Bolker2009.pdf)
- [Thiele & Markussen (2012)](https://www.uni-muenster.de/imperia/md/content/landschaftsoekologie/oekologischeplanung/team/publikationenjan/potential_of_glmm.pdf)

I also found this preprint, that may be worth looking into: [Meteyard & Davies (2019)](https://psyarxiv.com/h3duq/) with example tables in the [appendix (A5)](https://osf.io/6e8hu/).

**Comment by khusmann** [comment]
In the interim, I've started playing around with [kableExtra](http://haozhu233.github.io/kableExtra/awesome_table_in_html.html) which is worth a look. I'm having a lot of success rolling a latex lme4 results table by piping the results of the tidy output from [broom.mixed](https://cran.r-project.org/web/packages/broom.mixed/index.html).

It's also nice because it renders html when I'm running the chunk in rstudio, but automatically switches to latex when knitting. Super cool!

**Comment by crsh** [comment]
Thanks for the input. `papaja` uses `broom` to process model outputs and I was planning on using `broom.mixed`, too. It's good to know that it is useful for this purpose.

**Comment by crsh** [comment]
For a recent project, I've started on a first pass on this for fixed effects. This takes the input of a summary of `lmerTest `-output.

~~~r
library("dplyr")
library("papaja")
library("lmerTest")

lmer_test <- lmerTest::lmer(
  Reaction ~ Days + (Days | Subject)
  , data = sleepstudy
)

lmer_test_summary <- summary(lmer_test)

lmer_test_summary_table <- lmer_test_summary$coefficients %>%
  as.data.frame %>%
  rename(
    "$b$" = "Estimate"
    , "SE" = "Std. Error"
    , "$df$" = "df"
    , "$t$" = "t value"
    , "$p$" = "Pr(>|t|)"
  ) %>%
  mutate(
    Effect = papaja:::prettify_terms(rownames(.))
  ) %>%
  printnum(
    digits = c(2, 2, 2, 2, 3, 0)
    , gt1 = c(TRUE, TRUE, TRUE, TRUE, FALSE, FALSE)
    , zero = c(TRUE, TRUE, TRUE, TRUE, FALSE, FALSE)
  ) %>%
  select(Effect, `$b$`, `SE`, `$t$`, `$df$`, `$p$`)
~~~

The `apa_table()`-output should looks something like this:

<img width="654" alt="Bildschirmfoto 2019-07-08 um 15 41 34" src="https://user-images.githubusercontent.com/2195485/60815752-93e6fb00-a198-11e9-92a9-45154f2581b1.png">

Ideally, we would like a confidence interval instead of standard errors.

**Comment by crsh** [comment]
On information criteria Bolker et al. (2009):

> Variants of AIC are useful when sample sizes are small (AICc), when the data are overdispersed
(quasi-AIC, QAIC) or when one wants to identify the number of parameters in a ‘true’ model (Bayesian or Schwarz information criterion, BIC)

**Comment by nopsybot** [comment]
given that parameters and estimation procedures depend on the model estimated by lme4, the table could use automated footnotes (see the Meteyard & Davies guidance preprint appendix  https://osf.io/6e8hu/ page 24) "Table key"

**Comment by competulix** [comment]
Hey Frederik, what is the status on the mixed models reporting - this would be super helpful.

**Comment by crsh** [comment]
Hi Clemens, as far as I know, @mariusbarth has recently done some more work on this. I'm not sure what the current state of this is.

**Comment by mariusbarth** [comment]
Hi all, I created basic (and experimental) support for HLMs from lme4::`lmer()` and `lmerTest::lmer()` on my fork. You can install by typing `devtools::install_github("mariusbarth/papaja@devel")`.

There is a plethora of outputs from HLMs, but I suspect that only few of them are actually frequently used by papaja users. Therefore, I thought it may be worthwhile to ask: What are the output obejcts (created by which funtions) that should be necessarily be supported by `apa_print()`?

**Comment by crsh** [comment]
Great, could you briefly summarize what information you provide as of now?

**Comment by mcfrank** [comment]
I personally am generally interested in printing fixed effect coefficients
and fixed effect tables (I typically extract both from the `summary`
table).

Thanks Marius!

**Comment by mariusbarth** [comment]
For a model fitted with lme4, it looks like this:

```
> model_lme4 <- lme4::lmer(formula = yield ~ N + (1|block), data = npk)
> apa_print(model_lme4)
$estimate
$estimate$Intercept
[1] "$b = 52.07$, 95\\% CI $[48.17$, $55.97]$"

$estimate$N1
[1] "$b = 5.62$, 95\\% CI $[1.92$, $9.31]$"


$statistic
$statistic$Intercept
[1] "$t = 27.06$"

$statistic$N1
[1] "$t = 3.06$"


$full_result
$full_result$Intercept
[1] "$b = 52.07$, 95\\% CI $[48.17$, $55.97]$, $t = 27.06$"

$full_result$N1
[1] "$b = 5.62$, 95\\% CI $[1.92$, $9.31]$, $t = 3.06$"


$table
A data.frame with 4 labelled columns:

       term estimate       conf.int statistic
1 Intercept    52.07 [48.17, 55.97]     27.06
2        N1     5.62   [1.92, 9.31]      3.06

term     : Term 
estimate : $b$ 
conf.int : 95% CI 
statistic: $t$ 
attr(,"class")
[1] "apa_results" "list"   
```

------------------------
For a model fitted with lmerTest or converted to lmerModLmertest, it looks like this:
```
model_lmerTest <- lmerTest::as_lmerModLmerTest(model_lme4)
apa_print(model_lmerTest)

$estimate
$estimate$Intercept
[1] "$b = 52.07$, 95\\% CI $[48.17$, $55.97]$"

$estimate$N1
[1] "$b = 5.62$, 95\\% CI $[1.92$, $9.31]$"


$statistic
$statistic$Intercept
[1] "$t(8.17) = 27.06$, $p < .001$"

$statistic$N1
[1] "$t(17.00) = 3.06$, $p = .007$"


$full_result
$full_result$Intercept
[1] "$b = 52.07$, 95\\% CI $[48.17$, $55.97]$, $t(8.17) = 27.06$, $p < .001$"

$full_result$N1
[1] "$b = 5.62$, 95\\% CI $[1.92$, $9.31]$, $t(17.00) = 3.06$, $p = .007$"


$table
A data.frame with 6 labelled columns:

       term estimate       conf.int statistic    df p.value
1 Intercept    52.07 [48.17, 55.97]     27.06  8.17  < .001
2        N1     5.62   [1.92, 9.31]      3.06 17.00    .007

term     : Term 
estimate : $b$ 
conf.int : 95% CI 
statistic: $t$ 
df       : $\mathit{df}$ 
... (1 more label)
attr(,"class")
[1] "apa_results" "list"   
```

**Comment by pettitta** [comment]
I tried downloading @mariusbarth 's fork and using it for lmer output. Unfortunately, it's saying that lmer is not supported. Any idea why this might be?

**Comment by mariusbarth** [comment]
Hi @pettitta, I forgot to post here that lmer support moved to its own branch on my fork (I'm currently merging **a lot** of new stuff into the main repository). If you want to try lmer support, type `remotes::install_github("mariusbarth/papaja@merMod-support")`.

However, a cautionary note: The branch fell a bit back compared to the master branch in the main repository, which is why some bug fixes (especially with regard to parsing the YAML header) are not available there. Still, I just installed from the merMod-Branch and all unti tests and the template worked, so it may be worth trying. If you experience any problems, report back to me, then I will try to work on a solution asap.

**Comment by pettitta** [comment]
Thanks! I appreciate that. I'll go ahead and give it a try!

**Comment by mariusbarth** [comment]
Great! Also, if you think that we should include some other information in the output, feel free to ask for it. And one more thing: Currently, models fitted with `lme4::lmer()`, `lmerTest::lmer()` and `afex::mixed` should be supported.

**Comment by pettitta** [comment]
One thing I realized is that, when I run an `anova()` on an `lm()` model, it returns in `$full_results$var` not only the F and p values, but the effect size (e.g., "$F(1, 515) = 34.52$, $\\mathit{MSE} = 20.60$, $p < .001$"). When I run an `anova()` on an `lmer()` model, it only returns the F and p value (e.g., "$F(1, 93.43) = 69.34$, $p < .001$"). It would be great if we could have similar outputs for anovas on lmer as they are on lm!

**Comment by mariusbarth** [comment]
Very good point! I will look into the code and check whether the output from `anova(lmer(...))` contains the information we need to calculate the additional info

**Comment by shirdekel** [comment]
Hey @mariusbarth, thanks for the addition! A small thing I noticed was that for some reason, `apa_print()` doesn't work for models whose data I pass through a pipe. Both cases work fine for `lm()` models.

``` r
library(lme4)
#> Loading required package: Matrix
  library(magrittr)
  library(papaja)
  gm1 <- glmer(cbind(incidence, size - incidence) ~ period + (1 | herd),
               data = cbpp, family = binomial)
  
  apa_print(gm1) # works
#> Computing profile confidence intervals ...
#> $estimate
#> $estimate$Intercept
#> [1] "$\\hat{\\beta} = -1.40$, 95\\% CI $[-1.90$, $-0.95]$"
#> 
#> $estimate$period2
#> [1] "$\\hat{\\beta} = -0.99$, 95\\% CI $[-1.62$, $-0.41]$"
#> 
#> $estimate$period3
#> [1] "$\\hat{\\beta} = -1.13$, 95\\% CI $[-1.80$, $-0.51]$"
#> 
#> $estimate$period4
#> [1] "$\\hat{\\beta} = -1.58$, 95\\% CI $[-2.50$, $-0.80]$"
#> 
#> 
#> $statistic
#> $statistic$Intercept
#> [1] "$z = -6.05$, $p < .001$"
#> 
#> $statistic$period2
#> [1] "$z = -3.27$, $p = .001$"
#> 
#> $statistic$period3
#> [1] "$z = -3.49$, $p < .001$"
#> 
#> $statistic$period4
#> [1] "$z = -3.74$, $p < .001$"
#> 
#> 
#> $full_result
#> $full_result$Intercept
#> [1] "$\\hat{\\beta} = -1.40$, 95\\% CI $[-1.90$, $-0.95]$, $z = -6.05$, $p < .001$"
#> 
#> $full_result$period2
#> [1] "$\\hat{\\beta} = -0.99$, 95\\% CI $[-1.62$, $-0.41]$, $z = -3.27$, $p = .001$"
#> 
#> $full_result$period3
#> [1] "$\\hat{\\beta} = -1.13$, 95\\% CI $[-1.80$, $-0.51]$, $z = -3.49$, $p < .001$"
#> 
#> $full_result$period4
#> [1] "$\\hat{\\beta} = -1.58$, 95\\% CI $[-2.50$, $-0.80]$, $z = -3.74$, $p < .001$"
#> 
#> 
#> $table
#> A data.frame with 5 labelled columns:
#> 
#>        term estimate       conf.int statistic p.value
#> 1 Intercept    -1.40 [-1.90, -0.95]     -6.05  < .001
#> 2   Period2    -0.99 [-1.62, -0.41]     -3.27    .001
#> 3   Period3    -1.13 [-1.80, -0.51]     -3.49  < .001
#> 4   Period4    -1.58 [-2.50, -0.80]     -3.74  < .001
#> 
#> term     : Term 
#> estimate : $\hat{\beta}$ 
#> conf.int : 95\% CI 
#> statistic: $z$ 
#> p.value  : $p$ 
#> attr(,"class")
#> [1] "apa_results" "list"
  
  gm2 <- cbpp %>% 
    glmer(cbind(incidence, size - incidence) ~ period + (1 | herd),
          data = ., family = binomial)
  
  apa_print(gm2) # doesn't work
#> Computing profile confidence intervals ...
#> Error: 'data' not found, and some variables missing from formula environment
  
  lm1 <- lm(formula = mpg ~ factor(am), data = mtcars)
  
  apa_print(lm1) # works
#> $estimate
#> $estimate$Intercept
#> [1] "$b = 17.15$, 95\\% CI $[14.85$, $19.44]$"
#> 
#> $estimate$factoram1
#> [1] "$b = 7.24$, 95\\% CI $[3.64$, $10.85]$"
#> 
#> $estimate$modelfit
#> $estimate$modelfit$r2
#> [1] "$R^2 = .36$, 90\\% CI $[0.13$, $0.57]$"
#> 
#> $estimate$modelfit$r2_adj
#> [1] "$R^2_{adj} = .34$"
#> 
#> $estimate$modelfit$aic
#> [1] "$\\mathrm{AIC} = 196.48$"
#> 
#> $estimate$modelfit$bic
#> [1] "$\\mathrm{BIC} = 200.88$"
#> 
#> 
#> 
#> $statistic
#> $statistic$Intercept
#> [1] "$t(30) = 15.25$, $p < .001$"
#> 
#> $statistic$factoram1
#> [1] "$t(30) = 4.11$, $p < .001$"
#> 
#> $statistic$modelfit
#> $statistic$modelfit$r2
#> [1] "$F(1, 30) = 16.86$, $p < .001$"
#> 
#> 
#> 
#> $full_result
#> $full_result$Intercept
#> [1] "$b = 17.15$, 95\\% CI $[14.85$, $19.44]$, $t(30) = 15.25$, $p < .001$"
#> 
#> $full_result$factoram1
#> [1] "$b = 7.24$, 95\\% CI $[3.64$, $10.85]$, $t(30) = 4.11$, $p < .001$"
#> 
#> $full_result$modelfit
#> $full_result$modelfit$r2
#> [1] "$R^2 = .36$, 90\\% CI $[0.13$, $0.57]$, $F(1, 30) = 16.86$, $p < .001$"
#> 
#> 
#> 
#> $table
#> A data.frame with 5 labelled columns:
#> 
#>   predictor estimate                 ci statistic p.value
#> 1 Intercept    17.15 $[14.85$, $19.44]$     15.25  < .001
#> 2 Factoram1     7.24  $[3.64$, $10.85]$      4.11  < .001
#> 
#> predictor: Predictor 
#> estimate : $b$ 
#> ci       : 95\% CI 
#> statistic: $t(30)$ 
#> p.value  : $p$ 
#> attr(,"class")
#> [1] "apa_results" "list"
  
  lm2 <- mtcars %>% 
    lm(formula = mpg ~ factor(am), data = .)
  
  apa_print(lm2) # works
#> $estimate
#> $estimate$Intercept
#> [1] "$b = 17.15$, 95\\% CI $[14.85$, $19.44]$"
#> 
#> $estimate$factoram1
#> [1] "$b = 7.24$, 95\\% CI $[3.64$, $10.85]$"
#> 
#> $estimate$modelfit
#> $estimate$modelfit$r2
#> [1] "$R^2 = .36$, 90\\% CI $[0.13$, $0.57]$"
#> 
#> $estimate$modelfit$r2_adj
#> [1] "$R^2_{adj} = .34$"
#> 
#> $estimate$modelfit$aic
#> [1] "$\\mathrm{AIC} = 196.48$"
#> 
#> $estimate$modelfit$bic
#> [1] "$\\mathrm{BIC} = 200.88$"
#> 
#> 
#> 
#> $statistic
#> $statistic$Intercept
#> [1] "$t(30) = 15.25$, $p < .001$"
#> 
#> $statistic$factoram1
#> [1] "$t(30) = 4.11$, $p < .001$"
#> 
#> $statistic$modelfit
#> $statistic$modelfit$r2
#> [1] "$F(1, 30) = 16.86$, $p < .001$"
#> 
#> 
#> 
#> $full_result
#> $full_result$Intercept
#> [1] "$b = 17.15$, 95\\% CI $[14.85$, $19.44]$, $t(30) = 15.25$, $p < .001$"
#> 
#> $full_result$factoram1
#> [1] "$b = 7.24$, 95\\% CI $[3.64$, $10.85]$, $t(30) = 4.11$, $p < .001$"
#> 
#> $full_result$modelfit
#> $full_result$modelfit$r2
#> [1] "$R^2 = .36$, 90\\% CI $[0.13$, $0.57]$, $F(1, 30) = 16.86$, $p < .001$"
#> 
#> 
#> 
#> $table
#> A data.frame with 5 labelled columns:
#> 
#>   predictor estimate                 ci statistic p.value
#> 1 Intercept    17.15 $[14.85$, $19.44]$     15.25  < .001
#> 2 Factoram1     7.24  $[3.64$, $10.85]$      4.11  < .001
#> 
#> predictor: Predictor 
#> estimate : $b$ 
#> ci       : 95\% CI 
#> statistic: $t(30)$ 
#> p.value  : $p$ 
#> attr(,"class")
#> [1] "apa_results" "list"
```

<sup>Created on 2020-08-06 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

<details>

<summary>Session info</summary>

``` r
devtools::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.1 (2020-06-06)
#>  os       macOS Mojave 10.14.6        
#>  system   x86_64, darwin17.0          
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_AU.UTF-8                 
#>  ctype    en_AU.UTF-8                 
#>  tz       Australia/Sydney            
#>  date     2020-08-06                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version    date       lib source                             
#>  assertthat    0.2.1      2019-03-21 [1] CRAN (R 4.0.0)                     
#>  backports     1.1.8      2020-06-17 [1] CRAN (R 4.0.1)                     
#>  boot          1.3-25     2020-04-26 [1] CRAN (R 4.0.1)                     
#>  broom         0.7.0.9001 2020-08-06 [1] Github (tidymodels/broom@762e3ad)  
#>  callr         3.4.3      2020-03-28 [1] CRAN (R 4.0.0)                     
#>  cli           2.0.2      2020-02-28 [1] CRAN (R 4.0.0)                     
#>  crayon        1.3.4      2017-09-16 [1] CRAN (R 4.0.0)                     
#>  desc          1.2.0      2018-05-01 [1] CRAN (R 4.0.0)                     
#>  devtools      2.3.0      2020-04-10 [1] CRAN (R 4.0.0)                     
#>  digest        0.6.25     2020-02-23 [1] CRAN (R 4.0.0)                     
#>  dplyr         1.0.1      2020-07-31 [1] CRAN (R 4.0.2)                     
#>  ellipsis      0.3.1      2020-05-15 [1] CRAN (R 4.0.0)                     
#>  evaluate      0.14       2019-05-28 [1] CRAN (R 4.0.0)                     
#>  fansi         0.4.1      2020-01-08 [1] CRAN (R 4.0.0)                     
#>  fs            1.4.1      2020-04-04 [1] CRAN (R 4.0.0)                     
#>  generics      0.0.2      2018-11-29 [1] CRAN (R 4.0.0)                     
#>  glue          1.4.1      2020-05-13 [1] CRAN (R 4.0.0)                     
#>  highr         0.8        2019-03-20 [1] CRAN (R 4.0.0)                     
#>  htmltools     0.5.0      2020-06-16 [1] CRAN (R 4.0.1)                     
#>  knitr         1.29       2020-06-23 [1] CRAN (R 4.0.2)                     
#>  lattice       0.20-41    2020-04-02 [1] CRAN (R 4.0.1)                     
#>  lifecycle     0.2.0      2020-03-06 [1] CRAN (R 4.0.0)                     
#>  lme4        * 1.1-23     2020-04-07 [1] CRAN (R 4.0.0)                     
#>  magrittr    * 1.5        2014-11-22 [1] CRAN (R 4.0.0)                     
#>  MASS          7.3-51.6   2020-04-26 [1] CRAN (R 4.0.1)                     
#>  Matrix      * 1.2-18     2019-11-27 [1] CRAN (R 4.0.1)                     
#>  MBESS         4.7.0      2020-05-15 [1] CRAN (R 4.0.0)                     
#>  memoise       1.1.0      2017-04-21 [1] CRAN (R 4.0.0)                     
#>  minqa         1.2.4      2014-10-09 [1] CRAN (R 4.0.0)                     
#>  nlme          3.1-148    2020-05-24 [1] CRAN (R 4.0.1)                     
#>  nloptr        1.2.2.1    2020-03-11 [1] CRAN (R 4.0.0)                     
#>  papaja      * 0.1.0.9942 2020-08-06 [1] Github (mariusbarth/papaja@59edf05)
#>  pillar        1.4.6      2020-07-10 [1] CRAN (R 4.0.1)                     
#>  pkgbuild      1.0.8      2020-05-07 [1] CRAN (R 4.0.0)                     
#>  pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 4.0.0)                     
#>  pkgload       1.1.0      2020-05-29 [1] CRAN (R 4.0.0)                     
#>  prettyunits   1.1.1      2020-01-24 [1] CRAN (R 4.0.0)                     
#>  processx      3.4.2      2020-02-09 [1] CRAN (R 4.0.0)                     
#>  ps            1.3.3      2020-05-08 [1] CRAN (R 4.0.0)                     
#>  purrr         0.3.4      2020-04-17 [1] CRAN (R 4.0.0)                     
#>  R6            2.4.1      2019-11-12 [1] CRAN (R 4.0.0)                     
#>  Rcpp          1.0.5      2020-07-06 [1] CRAN (R 4.0.0)                     
#>  remotes       2.1.1      2020-02-15 [1] CRAN (R 4.0.0)                     
#>  rlang         0.4.7      2020-07-09 [1] CRAN (R 4.0.1)                     
#>  rmarkdown     2.3        2020-06-18 [1] CRAN (R 4.0.1)                     
#>  rprojroot     1.3-2      2018-01-03 [1] CRAN (R 4.0.0)                     
#>  sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 4.0.0)                     
#>  statmod       1.4.34     2020-02-17 [1] CRAN (R 4.0.0)                     
#>  stringi       1.4.6      2020-02-17 [1] CRAN (R 4.0.0)                     
#>  stringr       1.4.0      2019-02-10 [1] CRAN (R 4.0.0)                     
#>  testthat      2.3.2      2020-03-02 [1] CRAN (R 4.0.0)                     
#>  tibble        3.0.3      2020-07-10 [1] CRAN (R 4.0.1)                     
#>  tidyr         1.1.1      2020-07-31 [1] CRAN (R 4.0.2)                     
#>  tidyselect    1.1.0      2020-05-11 [1] CRAN (R 4.0.0)                     
#>  usethis       1.6.1      2020-04-29 [1] CRAN (R 4.0.0)                     
#>  vctrs         0.3.2      2020-07-15 [1] CRAN (R 4.0.2)                     
#>  withr         2.2.0      2020-04-20 [1] CRAN (R 4.0.0)                     
#>  xfun          0.16       2020-07-24 [1] CRAN (R 4.0.2)                     
#>  yaml          2.2.1      2020-02-01 [1] CRAN (R 4.0.0)                     
#> 
#> [1] /Library/Frameworks/R.framework/Versions/4.0/Resources/library
```

</details>

**Comment by mariusbarth** [comment]
Hey @shirdekel, thanks for reporting! I dug a bit into the code of **lme4** and found that the function that calculates *profile* confindence intervals (which we call from within `apa_print()`) does not play well with pipes. I think this needs to be solved upstream, and I will try to create an issue at **lme4**, soon. In the meantime, I switched the default computation method for confidence intervals to `"Wald"`, which should work for both piped and non-piped workflows (same seems to be true for bootstrap CIs). If you want to install the new default behavior, type
```
remotes::install_github("mariusbarth/papaja@devel")`
```
If you don't want to reinstall, you can also manually require Wald CIs:
```
gm2 <- cbpp %>% 
    glmer(cbind(incidence, size - incidence) ~ period + (1 | herd),
          data = ., family = binomial)

apa_print(gm2, args_confint = list(method = "Wald"))
```

**Comment by crsh** [comment]
Hi everyone, @mariusbarth has been making great progress on implementing this since we have recently [updated our internal workflow](https://github.com/crsh/papaja/pull/372). 🎉  See [this PR](https://github.com/crsh/papaja/pull/395) for the current draft.

We are now wondering about the best way to handle the different model components (i.e. fixed and random effects). For context, the standard structure of `apa_print()`-output is currently 4-element nested list. Consider the following non-hierarchical example:

~~~r
library("papaja")
lm_mod <- lm(Sepal.Length ~ Petal.Width + Petal.Length, data = iris)
str(apa_print(lm_mod), give.attr = FALSE, give.head = FALSE, vec.len = 0)
~~~
~~~
List of 4
 $ estimate   :List of 4
  ..$ Intercept   : ...
  ..$ Petal_Width : ...
  ..$ Petal_Length: ...
  ..$ modelfit    :List of 4
  .. ..$ r2    : ...
  .. ..$ r2_adj: ...
  .. ..$ aic   : ...
  .. ..$ bic   : ...
 $ statistic  :List of 4
  ..$ Intercept   : ...
  ..$ Petal_Width : ...
  ..$ Petal_Length: ...
  ..$ modelfit    :List of 1
  .. ..$ r2: ...
 $ full_result:List of 4
  ..$ Intercept   : ...
  ..$ Petal_Width : ...
  ..$ Petal_Length: ...
  ..$ modelfit    :List of 1
  .. ..$ r2: ...
 $ table      :Classes ‘apa_results_table’ and 'data.frame':	3 obs. of  5 variables:
  ..$ predictor: ...
  ..$ estimate : ...
  ..$ ci       : ...
  ..$ statistic: ...
  ..$ p.value  : ...
~~~

Conceptually, we summarize the information in a table and use that table to construct strings for in-text reporting.

Now, this work well if the model output can be summarized in one table (and maybe some additional bits, such as in this case modelfit-information), but it is not immediately clear how to best handle a situation when the analysis output becomes more complex. I'd be interested in everyone's thoughts on this. Using the example of hierarchical linear models, I currently see three options:

###  1. Combined table

Something along these lines:

![](https://www.researchgate.net/profile/Sven_Bacher/publication/5300915/figure/tbl1/AS:601679321890850@1520462937788/Fixed-and-random-effects-in-a-generalized-linear-mixed-model-explaining-variation-in_W640.jpg)

While this would be a concise way of summarizing a model, I see several problems:

- I don't see a good way to include random or fixed effect correlations.
- The table will quickly become undwieldy for more complex models.

###  2. Multiple tables

Another option would be to return multiple tables (e.g. one for fixed and one for random effects) as named list elements of `table`. This is conceptually the same thing we do with the other output elements, such as `estimate` and `statistic`. It could look something like this:

~~~
$ table   :List of 2
  $ fixed_effects      :Classes ‘apa_results_table’ and 'data.frame':	3 obs. of  5 variables:
    ..$ predictor: ...
    ..$ estimate : ...
    ..$ ci       : ...
    ..$ statistic: ...
    ..$ p.value  : ...
  $ random_effects      :Classes ‘apa_results_table’ and 'data.frame':	3 obs. of  5 variables:
    ...
~~~

This would be convenient, but would complicate the structure of the other list elements for models with random slopes. We could do something like this (or in reverse nesting order):

~~~
$ estimate   :List of 4
  ..$ fixed_effects         :List of 3
    ..$ Intercept   : ...
    ..$ Petal_Width : ...
    ..$ Petal_Length: ...
  ..$ random_effects    :List of 3
    ..$ Intercept   : ...
    ..$ Petal_Width : ...
    ..$ Petal_Length: ...
~~~

This solution would be convenient in the sense that one call to `apa_print()` would return everything, but it would also result in a rather complex and hard to anticipate output structure (especially in the larger context of other `apa_print()`-methods).

### 3. Conditional tables

Finally, we could mirror the behavior of `broom.mixed::tidy()` and add an argument that specifies the type of information to return (e.g. `effects = c("fixed", "random")`). For full reporting this would require multiple calls to `apa_print()` but would ensure a consistent output structure (if this becomes an issue, I could imagine allowing users to pass a vector and return a list with two elements containing the separate outputs for fixed and random effects).

I currently tend towards this approach, but again I'd be interested in everyone's thoughts on this.

**Comment by pettitta** [comment]
As I was reading through your choices, I got to choice 2 and I thought to myself "it would be nice if we could just *choose* which parts we want to use so that there is some flexibility. And then I got to choice 3 and you had already considered it! I think that's the way to go, imo.

**Comment by jstevens5** [comment]
I agree. The third option is the most flexible.
Great job, you all!

**Comment by crsh** [comment]
Some more recommendations here: https://doi.org/10.1016/j.jml.2020.104092 https://osf.io/bfq39/

**Comment by Misch007** [comment]
Hey!
I am very new to R, GitHub and programming in general, so sorry in advance if this issue doesn't belong here or has been addressed somewhere else already or something's wrong with the format!

I am using the following code to compute a GLMM:
```
remotes::install_github("crsh/papaja@devel")
library(papaja)
library(afex) 

glmm <- mixed(Errors ~ Group*Session*Task+(1|Participant), df, family='poisson', method='LRT')
```

When I try to use 

`apa_print(glmm)` 

it gives me the following error message:

`Error in as.character(x) : 
  cannot coerce type 'closure' to vector of type 'character'`

It does give me an output, though, if I use 

`apa_print(glmm$full_model)`, 

the one thing I do not get then is the ANOVA table, but instead the table I'd receive with `summary(glmm)` . It must have something to do with the `family='poisson, method="LRT"` subset, because it works fine without it. Interestingly, if I leave out the subset...

`glmm <- mixed(Errors ~ Group*Session*Task+(1|Participant), df)` 

...and then try again to use... 

`apa_print(glmm$full_model),`

...it gives me the same error, which I got **with** the subset and trying to print the **without** the `$full_model` specification:

`Error in as.character(x) : 
  cannot coerce type 'closure' to vector of type 'character'`

Let me know, if you need more information and thanks for the great work on this one!

**Comment by mariusbarth** [comment]
Hi @Misch007, and welcome to the wonderful world of R stats!

Unfortunately, I cannot reproduce the problem you are experiencing with the information you already provided. Could you also provide a) an overview of the contents of `df` (or, if possible, you try to create an example with one of R's built-in data sets) and b) the output of `sessionInfo()`

**Comment by Misch007** [comment]
Hey @mariusbarth, thanks for the warm welcoming!

So I recreated my data manually (sorry, if it isn't the straight forwards way) and again get the same error message:

```
afex_options(emmeans_model = "multivariate")

participant<-c("vp1","vp2","vp3","vp4","vp5")
group<-c("intervention", "control","control","intervention","control")

df<-data.frame(participant,group) %>% 
  group_by(participant,group) %>% 
  summarise(session=c("t1","t2","t3","t4")) %>% 
  group_by(participant, group, session) %>% 
  summarise(task=c("a","b")) %>% 
  ungroup() %>% 
  mutate(errors=floor(runif(n=40,min=0,max=30)))

glmm <- mixed(errors~group*session*task+(1|participant), df, family='poisson', method='LRT')

apa_print(glmm)

Error in as.character(x) : 
  cannot coerce type 'closure' to vector of type 'character'
```

the output of `sessionInfo()` looks as follows
```
> sessionInfo()
R version 4.1.0 (2021-05-18)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19043)

Matrix products: default

locale:
[1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252    LC_MONETARY=German_Germany.1252
[4] LC_NUMERIC=C                    LC_TIME=German_Germany.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] MOTE_1.0.2         ggbeeswarm_0.6.0   forcats_0.5.1      stringr_1.4.0      dplyr_1.0.7        purrr_0.3.4       
 [7] readr_1.4.0        tidyr_1.1.3        tibble_3.1.3       tidyverse_1.3.1    rmarkdown_2.9      devtools_2.4.2    
[13] usethis_2.0.1      ggstance_0.3.5     RColorBrewer_1.1-2 gmodels_2.18.1     ggrepel_0.9.1      ggpubr_0.4.0      
[19] ggplot2_3.3.5      coin_1.4-1         survival_3.2-11    psych_2.1.6        rstatix_0.7.0      ggstatsplot_0.8.0 
[25] summarytools_0.9.9 car_3.0-11         carData_3.0-4      knitr_1.33         emmeans_1.6.2-1    afex_0.28-1       
[31] lme4_1.1-27.1      Matrix_1.3-3       dtplyr_1.1.0       data.table_1.14.0  openxlsx_4.2.4     papaja_0.1.0.9997 
[37] tinylabels_0.2.1   pubprint_0.2.1    

loaded via a namespace (and not attached):
  [1] utf8_1.2.2                tidyselect_1.1.1          grid_4.1.0                gmp_0.6-2                
  [5] munsell_0.5.0             codetools_0.2-18          effectsize_0.4.5          withr_2.4.2              
  [9] colorspace_2.0-2          highr_0.9                 rstudioapi_0.13           stats4_4.1.0             
 [13] ipmisc_6.0.2              ggsignif_0.6.2            mnormt_2.0.2              rprojroot_2.0.2          
 [17] coda_0.19-4               vctrs_0.3.8               generics_0.1.0            TH.data_1.0-10           
 [21] xfun_0.24                 BWStest_0.2.2             R6_2.5.0                  BayesFactor_0.9.12-4.2   
 [25] cachem_1.0.5              reshape_0.8.8             assertthat_0.2.1          scales_1.1.1             
 [29] multcomp_1.4-17           beeswarm_0.4.0            gtable_0.3.0              multcompView_0.1-8       
 [33] processx_3.5.2            sandwich_3.0-1            rlang_0.4.11              MatrixModels_0.5-0       
 [37] zeallot_0.1.0             PMCMRplus_1.9.0           splines_4.1.0             broom_0.7.9.9000         
 [41] rapportools_1.0           checkmate_2.0.0           yaml_2.2.1                reshape2_1.4.4           
 [45] abind_1.4-5               modelr_0.1.8              backports_1.2.1           tools_4.1.0              
 [49] tcltk_4.1.0               ellipsis_0.3.2            WRS2_1.1-2                sessioninfo_1.1.1        
 [53] ez_4.4-0                  Rcpp_1.0.7                plyr_1.8.6                base64enc_0.1-3          
 [57] ps_1.6.0                  prettyunits_1.1.1         pbapply_1.4-3             correlation_0.6.1        
 [61] zoo_1.8-9                 haven_2.4.1               fs_1.5.0                  magrittr_2.0.1           
 [65] magick_2.7.2              lmerTest_3.1-3            reprex_2.0.0              tmvnsim_1.0-2            
 [69] mvtnorm_1.1-2             matrixStats_0.59.0        pkgload_1.2.1             hms_1.1.0                
 [73] patchwork_1.1.1           evaluate_0.14             xtable_1.8-4              rio_0.5.27               
 [77] pairwiseComparisons_3.1.6 readxl_1.3.1              testthat_3.0.4            compiler_4.1.0           
 [81] crayon_1.4.1              minqa_1.2.4               htmltools_0.5.1.1         mgcv_1.8-35              
 [85] mc2d_0.1-21               libcoin_1.0-8             MBESS_4.8.0               lubridate_1.7.10         
 [89] DBI_1.1.1                 SuppDists_1.1-9.5         kSamples_1.2-9            dbplyr_2.1.1             
 [93] MASS_7.3-54               boot_1.3-28               cli_3.0.1                 pryr_0.1.4               
 [97] gdata_2.18.0              parallel_4.1.0            insight_0.14.2            pkgconfig_2.0.3          
[101] statsExpressions_1.1.0    numDeriv_2016.8-1.1       foreign_0.8-81            xml2_1.3.2               
[105] paletteer_1.3.0           vipor_0.4.5               estimability_1.3          rvest_1.0.0              
[109] callr_3.7.0               digest_0.6.27             parameters_0.14.0         cellranger_1.1.0         
[113] curl_4.3.2                gtools_3.9.2              modeltools_0.2-23         nloptr_1.2.2.2           
[117] lifecycle_1.0.0           nlme_3.1-152              jsonlite_1.7.2            desc_1.3.0               
[121] fansi_0.5.0               pillar_1.6.2              lattice_0.20-44           fastmap_1.1.0            
[125] httr_1.4.2                pkgbuild_1.2.0            glue_1.4.2                remotes_2.4.0            
[129] bayestestR_0.10.0         zip_2.2.0                 pander_0.6.4              stringi_1.7.3            
[133] performance_0.7.2         rematch2_2.1.2            memoise_2.0.0             Rmpfr_0.8-4   
```
Thanks a lot for the help in advance!

**Comment by crsh** [comment]
Thanks for the reproducible example. I can reproduce the error on my end.

**Comment by mariusbarth** [comment]
Same here. The problem seems to be related to `glue_apa_results()`, really down the stream. @crsh, can you have a look at this one?  Here is a more isolated code bit that reproduces the error:

```
x <- structure(
  list(
    term = structure(c("Group", "Session", "Task"), label = "Effect", class = c("tiny_labelled", "character"))
    , statistic = structure(c("1.26", "6.25", "13.04"), label = "$\\chi^2$", class = c("tiny_labelled", "character"))
    , df = structure(c("1", "3", "1"), label = "$\\mathit{df}$", class = c("tiny_labelled", "character"))
    , p.value = structure(c(".262", ".100", "< .001"), label = "$p$", class = c("tiny_labelled", "character"))
  )
  , row.names = c(NA, -3L)
  , class = c("apa_results_table", "data.frame")
)

glue_apa_results(
  x
  , est_glue = papaja:::construct_glue(x, "estimate")
  , stat_glue = papaja:::construct_glue(x, "statistic")
)
```

**Comment by crsh** [comment]
The problem should be resolved on the development branch. Let me know if this resolves the issue on your end when it's convenient for you.

**Comment by Misch007** [comment]
Hey @crsh and @mariusbarth,

thanks a lot for the very quick help, it works perfectly on my end, now and I am very happy to continue writing using your package!

Best regards,
Mischa

**Comment by crsh** [comment]
Glad to hear it. Happy writing!

**Comment by raffelectrified** [comment]
dear all,

first of all, thanks @crsh and @mariusbarth for this incredible content and for papaja - I absolutely love it!!! Only recently have I used apa_print to report output from LMER and GLMER (using the lme4 package). What I find, however, is that it takes forever to compute profile confidence intervals to the point that it crashes during the computation of the second model (GLMER) and would not knit my Rmarkdown document. Overall, both models compute relatively quickly and are not overparameterized so I guess this should not be a problem with the models structure? 

The problem emerges when I call 

`rt_model_output <- apa_print(parsimonious_lmer)`
and
`acc_model_output <- apa_print(parsimonious_glmer)`

and the message I get looks the following:
`Computing profile confidence intervals . . .`

Has it happened to someone else before? I could also pull out the contents of the model with parsimonious_lmer$coefficients and some supplementary functions, but this does not compute as well as apa_print when I knit the document.

Thanks for any insights,
Best,
Raf.

**Comment by mariusbarth** [comment]
Hi @raffelectrified, you probably installed **papaja** from the master branch, right? We switched the defaults for calculating confidence intervals from profile to Wald, which is why you should be able to use `apa_print()` after installing from the development branch (via `remotes::install_github("crsh/papaja@devel"`). Can you check if this is indeed the case? Moreover, can you share the code that created the `glmer` model? I'm curious if I missed an edge case. 

Thanks in advance,
best regards

Marius

**Comment by raffelectrified** [comment]
Hi @mariusbarth, you're right, I did! Having installed the development branch the issue disappeared. Thank you so much! Also, the part of the code that computes GLMER and the apa_print() output from GLMER is the following:

`parsimonious_glmer <- glmer(WaitWord.ACC.Trial.~negative_vs_neutral*polish_vs_english*new_vs_old + (negative_vs_neutral*polish_vs_english*new_vs_old - negative_vs_neutral:polish_vs_english - negative_vs_neutral||Subject) + (1|Stimulus.Trial.), family="binomial", data = decode_acc_lmem, glmerControl(calc.derivs=FALSE))`

`summary(parsimonious_glmer)`
`VarCorr(parsimonious_glmer)`
`acc_model_output <- apa_print(parsimonious_glmer)`

**Comment by mariusbarth** [comment]
Hi @raffelectrified, great to hear! Just so that you know, `apa_print()` will now calculate "Wald" confidence intervals (see `?lme4::confint.merMod` for details). You can ask `apa_print()` to calculate profile CIs (for linear mixed models) or bootstrap CIs (for both linear and generalized linear models) via parameter `args_confint` (see the `?apa_print.merMod` for details).

**Comment by annalibey** [comment]
Hi all, I'm not sure if I'm misinterpreting above - does the development branch now support linear mixed models with random effects or only generalized hierarchical models? I'm using the lme4 package like so

`hwise_lm <- lmer(hwise_scale_score ~ jmp_level + gender + hh_income  + (1|cluster), data = model_data)`

`apa_lm <- apa_print(hwise_lm)`

`apa_table( apa_lm$table)`

When I get this error.
> Error in no_method(x) : 
  Objects of class 'lmerModLmerTest' are currently not supported (no method defined).
              Visit https://github.com/crsh/papaja/issues to request support for this class

**Comment by mariusbarth** [comment]
Hi @annalibey,

after installing from the development branch (via `remotes::install_github("crsh/papaja@devel"), both types of models should be supported.

**Comment by annalibey** [comment]
@mariusbarth thanks for the reply - it was an install error on my end and restarting RStudio fixed it. Great tool!

**Comment by mdkraemer** [comment]
First, thanks for all the work that went into this package! I'm currently preparing resubmitting a manuscript using papaja with lmer models and want to further improve computational reproducibility. I've written a short "Instructions to reproduce" document on how to install the development branches, and use the 'groundhog' package for the CRAN packages. Do you know any way to make sure that future attempts to reproduce the manuscript will refer to the same version of papaja? Currently, I include the following commands in my instructions to install papaja:  
`if(!"devtools" %in% rownames(installed.packages())) install.packages("devtools")`  
`devtools::install_github("crsh/papaja")`  
`remotes::install_github("mariusbarth/papaja@devel")`

**Comment by crsh** [comment]
Hi Michael,

thanks for reaching out. First, the prefered way to install `papaja` (including `lmer` support) is

~~~r
remotes::install_github("crsh/papaja@devel")
~~~

This will install the latest development version. To ensure that you always install the same version of the package, reference the exact commit you want to install. Currently, the latest commit on `devel` is `88fa35003f1b45b1cab4cb6a410f69689c4485e7`. If this is the version you would like to use install via

~~~r
remotes::install_github("crsh/papaja@88fa35003f1b45b1cab4cb6a410f69689c4485e7")
~~~

That said, I'm would recommend using Docker, for example, using the workflow I have outlined [here](https://github.com/crsh/papaja_docker). But I understand that this may not be for everyone.

**Comment by mcfrank** [comment]
I just wanted to come back to this (now very old) thread that I initiated and say thank you @crsh!!! `apa_print` is working perfectly for my use case, and it is incredibly easy to use. Thanks again for all your hard work on this awesome package - it's no exaggeration to say that it has radically altered the way that my lab and I work. 🎉🎉🎉

**Comment by crsh** [comment]
Hi Mike, thanks a lot for the kind words. This means a lot to us! @mariusbarth did most of the heavy lifting on the `lmer`-support. 🏅

**Comment by mariusbarth** [comment]
Jup, thanks fot the kind words!

**Comment by crsh** [comment]
I'm closing this issue as the feature has been released to the main branch. Any requests for further features should probably go into a new issue.
